db/src/main/java/org/odk/collect/db/sqlite/RowNumbers.kt

```diff
@@ -0,0 +1,63 @@
+package org.odk.collect.db.sqlite
+
+import android.database.Cursor
+import org.odk.collect.db.sqlite.SQLiteColumns.ROW_ID
+import org.odk.collect.db.sqlite.SQLiteColumns.ROW_NUMBER
+
+object RowNumbers {
+    fun SynchronizedDatabaseConnection.rawQueryWithRowNumber(table: String, selection: String? = null, selectionArgs: Array<String>? = null): Cursor {
+        ensureRowIdTable(this, table)
+
+        val cursor = if (selection != null) {
+            this.withConnection {
+                readableDatabase
+                    .rawQuery(
+                        """
+                        SELECT *, i.$ROW_ID as $ROW_NUMBER
+                        FROM "$table" e, "${getRowIdTableName(table)}" i
+                        WHERE e._id = i._id AND $selection
+                        ORDER BY i.$ROW_ID
+                        """.trimIndent(),
+                        selectionArgs
+                    )
+            }
+        } else {
+            this.withConnection {
+                readableDatabase
+                    .rawQuery(
+                        """
+                        SELECT *, i.$ROW_ID as $ROW_NUMBER
+                        FROM "$table" e, "${getRowIdTableName(table)}" i
+                        WHERE e._id = i._id
+                        ORDER BY i.$ROW_ID
+                        """.trimIndent(),
+                        null
+                    )
+            }
+        }
+
+        return cursor
+    }
+
+    fun SynchronizedDatabaseConnection.invalidateRowNumbers(table: String) {
```

Comment:
This function just drops the table so invalidateRowNumbers is a little bit misleading. A better name would be eg `dropRowIdTable`.

Reply:
I preferred the idea of keeping expressing it as "invalidate" to make it less implementation specific. Clients are just signalling that row numbers will change. I'm expecting that we'll end up using `row_number()` for newer OS versions soon enough and that this method will a no-op for OS version 30+.